calyx/opt/src/passes/profiler_instrumentation.rs

```diff
@@ -417,6 +508,97 @@ fn create_assignments<T: Clone>(
     group_name_assign_and_cell
 }
 
+fn continuous_assignments(
+    comp: &mut ir::Component,
+    sigs: &ir::LibrarySignatures,
+) {
+    // vector of cells (non-primitives) invoked
+    let mut cell_invoke_vec: Vec<(Id, ir::Guard<Nothing>)> = Vec::new();
+    // vector of primitives invoked
+    let mut primitive_invoke_vec: Vec<(Id, ir::Guard<Nothing>)> = Vec::new();
+
+    // set to prevent adding multiple probes for a combinational primitive
+    let mut comb_primitives_covered = HashSet::new();
+    let mut comb_cells_covered = HashSet::new();
+    for assignment_ref in comp.continuous_assignments.iter() {
+        let dst_borrow = assignment_ref.dst.borrow();
+        let guard = *(assignment_ref.guard.clone());
+        if let ir::PortParent::Cell(cell_ref) = &dst_borrow.parent {
+            match cell_ref.upgrade().borrow().prototype.clone() {
+                calyx_ir::CellType::Primitive {
+                    name: _,
+                    param_binding: _,
+                    is_comb: _,
+                    latency: _,
+                } => {
+                    let cell_name = cell_ref.upgrade().borrow().name();
+                    // collecting primitives for area utilization; we want to avoid adding the same primitive twice!
+                    if comb_primitives_covered.insert(cell_name) {
+                        primitive_invoke_vec.push((cell_name, guard));
+                    }
+                }
+                calyx_ir::CellType::Component { name: _ } => {
+                    let cell_name = cell_ref.upgrade().borrow().name();
+                    if comb_cells_covered.insert(cell_name) {
+                        cell_invoke_vec.push((cell_name, guard));
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+
+    // add probes for primitives in continuous assignment
+    let delimiter = "___";
+    let comp_name = comp.name;
+    let mut builder = ir::Builder::new(comp, sigs);
+    let one = builder.add_constant(1, 1);
+    let mut assign_and_cell = Vec::new();
+    for (primitive_cell_name, guard) in primitive_invoke_vec.iter() {
+        let probe_cell_name = format!(
+            "{}{}{}_contprimitive_probe",
+            primitive_cell_name, delimiter, comp_name
+        );
+        let probe_cell =
+            builder.add_primitive(probe_cell_name, "std_wire", &[1]);
+        probe_cell.borrow_mut().add_attribute(BoolAttr::Control, 1);
+        probe_cell
+            .borrow_mut()
+            .add_attribute(BoolAttr::Protected, 1);
+        let probe_asgn: ir::Assignment<Nothing> = builder.build_assignment(
+            probe_cell.borrow().get("in"),
+            one.borrow().get("out"),
+            guard.clone(),
+        );
+        assign_and_cell.push((probe_asgn, probe_cell));
+    }
+    // add probes for cells (non-primitives) in continuous assignment
+    for (primitive_cell_name, guard) in cell_invoke_vec.iter() {
```

Comment:
Let's rename the variable `primitive_cell_name` since we're not dealing with primitives here?